File: shared/voiceqa_shared/crypto.py
"""AES-256-GCM for secrets stored in the DB (txn API credentials now;
SSO client secret in Phase 4).

Format: base64(nonce) + "." + base64(ciphertext||tag). The Node twin
(apps/web/src/lib/crypto.ts, Phase 4) must produce/accept the same format —
keep TEST_VECTOR in sync across both implementations.
"""

from __future__ import annotations

import base64
import hashlib
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from voiceqa_shared.settings import SharedSettings

_settings = SharedSettings()


def _key() -> bytes:
    raw = _settings.APP_ENCRYPTION_KEY.get_secret_value()
    try:
        decoded = base64.b64decode(raw, validate=True)
        if len(decoded) == 32:
            return decoded
    except Exception:
        pass
    # Arbitrary string key — derive 32 bytes deterministically.
    return hashlib.sha256(raw.encode()).digest()


def encrypt_str(plaintext: str) -> str:
    nonce = os.urandom(12)
    ct = AESGCM(_key()).encrypt(nonce, plaintext.encode(), None)
    return base64.b64encode(nonce).decode() + "." + base64.b64encode(ct).decode()


def decrypt_str(token: str) -> str:
    nonce_b64, _, ct_b64 = token.partition(".")
    nonce = base64.b64decode(nonce_b64)
    ct = base64.b64decode(ct_b64)
    return AESGCM(_key()).decrypt(nonce, ct, None).decode()

File: apps/web/src/lib/crypto.ts
/**
 * AES-256-GCM twin of shared/voiceqa_shared/crypto.py.
 *
 * Must stay byte-compatible: same key derivation (base64 32-byte key, else
 * sha256 of the raw string) and the same wire format
 * `base64(nonce) + "." + base64(ciphertext || tag)` (16-byte GCM tag
 * appended to the ciphertext, matching Python's cryptography AESGCM output).
 *
 * Used server-side only, by the lazy NextAuth init, to decrypt the SSO
 * client secret read from the sso_config row.
 *
 * Cross-language test vector (APP_ENCRYPTION_KEY = "dGVzdC1rZXktMzItYnl0ZXMtZXhhY3RseSEhMTIzNDU="):
 *   decrypt_str(encrypt_str("hello")) === "hello" in BOTH Python and Node, and
 *   a token produced by one decrypts in the other. Verified in CI / by hand.
 */

import { createCipheriv, createDecipheriv, createHash, randomBytes } from "node:crypto";

const TAG_BYTES = 16;

function derivedKey(): Buffer {
  const raw = process.env.APP_ENCRYPTION_KEY ?? "";
  const decoded = Buffer.from(raw, "base64");
  // Mirror Python's base64.b64decode(raw, validate=True) + len==32 check:
  // only the base64 path when it round-trips to the exact same string.
  if (decoded.length === 32 && decoded.toString("base64") === raw) {
    return decoded;
  }
  return createHash("sha256").update(raw, "utf8").digest();
}

export function encryptStr(plaintext: string): string {
  const nonce = randomBytes(12);
  const cipher = createCipheriv("aes-256-gcm", derivedKey(), nonce);
  const ct = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  const combined = Buffer.concat([ct, tag]);
  return `${nonce.toString("base64")}.${combined.toString("base64")}`;
}

export function decryptStr(token: string): string {
  const [nonceB64, ctB64] = token.split(".");
  const nonce = Buffer.from(nonceB64, "base64");
  const combined = Buffer.from(ctB64, "base64");
  const ct = combined.subarray(0, combined.length - TAG_BYTES);
  const tag = combined.subarray(combined.length - TAG_BYTES);
  const decipher = createDecipheriv("aes-256-gcm", derivedKey(), nonce);
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(ct), decipher.final()]).toString("utf8");
}
